### ISSUE TITLE
paramiko + pynacl in django project delays manage.py shell

### ISSUE THREAD

**Issue by bastbnl** [paramiko/paramiko#1023]
I'd like to thank you for building and maintaining this package. I've used it in all kinds of projects and I really appreciate its flexibility.  Unfortunately, I ran into an issue that took some time to resolve.

So here's a summary. I'm building a Django project and I'm using Fabric within that project. The project uses a virtualenv. I'm using `python manage.py shell` quite regularly and I noticed starting the shell is relatively fast and it takes far too much time every now and then to get access to the shell. I used `strace` to see what's going on and here's a fragment of what I found:

```
21:03:22 close(17)                      = 0
21:03:22 close(16)                      = 0
21:03:22 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core", 0x7ffe849331a0) = -1 ENOENT
 (No such file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.x86_64-linux-gnu.so", O_RDONL
Y) = -1 ENOENT (No such file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.so", O_RDONLY) = -1 ENOENT (N
o such file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_coremodule.so", O_RDONLY) = -1 ENO
ENT (No such file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.py", O_RDONLY) = 16
21:03:22 fstat(16, {st_mode=S_IFREG|0644, st_size=1079, ...}) = 0
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.pyc", O_RDONLY) = 17
21:03:22 fstat(17, {st_mode=S_IFREG|0644, st_size=1047, ...}) = 0
21:03:22 read(17, "\3\363\r\nf\345_Yc\0\0\0\0\0\0\0\0\2\0\0\0@`\1\0sh\0\0\0d\0"..., 4096) = 1047
21:03:22 fstat(17, {st_mode=S_IFREG|0644, st_size=1047, ...}) = 0
21:03:22 read(17, "", 4096)             = 0
21:03:22 close(17)                      = 0
21:03:22 close(16)                      = 0
21:03:22 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils", 0x7ffe849331a0) = -1 ENOENT (No s
uch file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.x86_64-linux-gnu.so", O_RDONLY) = -
1 ENOENT (No such file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.so", O_RDONLY) = -1 ENOENT (No such
 file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utilsmodule.so", O_RDONLY) = -1 ENOENT (N
o such file or directory)
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.py", O_RDONLY) = 16
21:03:22 fstat(16, {st_mode=S_IFREG|0644, st_size=1311, ...}) = 0
21:03:22 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.pyc", O_RDONLY) = 17
21:03:22 fstat(17, {st_mode=S_IFREG|0644, st_size=1087, ...}) = 0
21:03:22 read(17, "\3\363\r\nf\345_Yc\0\0\0\0\0\0\0\0\2\0\0\0@`\1\0s^\0\0\0d\0"..., 4096) = 1087
21:03:22 fstat(17, {st_mode=S_IFREG|0644, st_size=1087, ...}) = 0
21:03:22 read(17, "", 4096)             = 0
21:03:22 close(17)                      = 0
21:03:22 close(16)                      = 0
21:03:22 open("/dev/random", O_RDONLY)  = 16
21:03:22 poll([{fd=16, events=POLLIN}], 1, -1) = 1 ([{fd=16, revents=POLLIN}])
21:04:35 close(16)                      = 0
21:04:35 open("/dev/urandom", O_RDONLY) = 16
21:04:35 fstat(16, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 9), ...}) = 0
21:04:35 fcntl(16, F_GETFD)             = 0
21:04:35 fcntl(16, F_SETFD, FD_CLOEXEC) = 0
21:04:35 read(16, "\367\17\215\375\23\313\217\26N\21Qf\252\320>\302", 16) = 16
21:04:35 close(15)                      = 0
21:04:35 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/encoding", 0x7ffe84933700) = -1 ENOENT (No such file or directory)
```

It looks like the nacl module is waiting for content from `/dev/random`. That probably means that I need to inspect the system configuration somewhere, but I wasn't able to find some quick pointers. I rolled back to paramiko==2.1.3 and I'm not bothered by this issue anymore.

Looking for a way to reproduce this issue? Use a single core 1GB VPS (yes...) on Debian 9.1, setup a virtualenv using python 2.7, include paramiko and Django 1.11.3 and start `python manage.py shell` for a number of times. It'll probably take you about 26 attempts (maybe even more) to run into the issue. I've seen it happen faster, but I don't want to get your hopes up.

**Comment by ploxiln** [comment]
I think you could trigger the issue reliably by "draining" `/dev/random` e.g. with "cat </dev/random >/dev/null".

The nacl module ought to be using `/dev/urandom` from the start instead of trying `/dev/random` first, IMHO. But also, we should see if its initializing itself before it really needs to...

**Comment by ploxiln** [comment]
I think that poll is this here: https://github.com/pyca/pynacl/blob/1.1.2/src/libsodium/src/libsodium/randombytes/sysrandom/randombytes_sysrandom.c#L133

So what that does is just guarantee that the entropy pool is initialized, by seeing if it *could* read from `/dev/random` - it does not actually read from it. (Otherwise, if it's immediately after bootup, it's possible that it could get bad entropy from `/dev/urandom` because the entropy pool was not fully initialized.) But this method can't tell when a server had urandom fully initialized and then had it's `/dev/random` entropy estimate depleted. (But the whole concept of "being depleted" is a silly.)

Anyway, since it seems like your use case doesn't actually need any of the pynacl functionality, it does seem like it would be nice to avoid needing to initialize it.

cc @alex

**Comment by alex** [comment]
I don't even think FIPS requires this "depleted nonsense" (and I ordinarily have _nothing_ nice to say about FIPS).

paramiko has no control over this, so we should probably move this to a different issue tracker.

What kernel version are you running? Probably the easiest way to resolve this is to use a kernel with `getrandom`, which will avoid the need to much with this stuff.

/cc @reaperhulk

**Comment by bastbnl** [comment]
Wow... that's fast. 

@alex  I'm using this
```
bastb@v:~$ uname -a
Linux v 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u2 (2017-06-26) x86_64 GNU/Linux

bastb@v:~$ grep -i random /boot/config-4.9.0-3-amd64
CONFIG_SLAB_FREELIST_RANDOM=y
CONFIG_ARCH_HAS_ELF_RANDOMIZE=y
CONFIG_ARCH_RANDOM=y
CONFIG_RANDOMIZE_BASE=y
CONFIG_RANDOMIZE_MEMORY=y
CONFIG_RANDOMIZE_MEMORY_PHYSICAL_PADDING=0xa
CONFIG_NET_TEAM_MODE_RANDOM=m
CONFIG_HW_RANDOM=m
# CONFIG_HW_RANDOM_TIMERIOMEM is not set
CONFIG_HW_RANDOM_INTEL=m
CONFIG_HW_RANDOM_AMD=m
CONFIG_HW_RANDOM_VIA=m
CONFIG_HW_RANDOM_VIRTIO=m
CONFIG_HW_RANDOM_TPM=m
# Random Number Generation
# CONFIG_RANDOM32_SELFTEST is not set

```

**Comment by reaperhulk** [comment]
Do you see this problem if you reinstall pynacl without using the manylinux1 wheel? That can be accomplished by adding --no-binary pynacl to the pip install command. I suspect the problem here is that the manylinux1 compilation of libsodium is using the /dev/urandom path and libsodium doesn't do runtime detection.

**Comment by bastbnl** [comment]
**TL;DR**: The `--no-binary pynacl ` option worked. Thanks for your help.

@reaperhulk I installed that version using this command line:
```
(env-brownpapersession) bastb@bastb-vps:/var/www/brownpapersession/dev/brownpapersession$ pip install pynacl -I --no-binary pynacl
Collecting pynacl
  Using cached PyNaCl-1.1.2.tar.gz
Collecting six (from pynacl)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting cffi>=1.4.1 (from pynacl)
  Using cached cffi-1.10.0-cp27-cp27mu-manylinux1_x86_64.whl
Collecting pycparser (from cffi>=1.4.1->pynacl)
Skipping bdist_wheel for pynacl, due to binaries being disabled for it.
Installing collected packages: six, pycparser, cffi, pynacl
  Running setup.py install for pynacl ... done
Successfully installed cffi-1.10.0 pycparser-2.18 pynacl-1.1.2 six-1.10.0
Segmentation fault
```

Something is causing a Segmentation fault, but it does not appear to be `nacl` as that's installed:
```
(env-brownpapersession) bastb@bastb-vps:/var/www/brownpapersession/dev/brownpapersession$ python manage.py shell
Python 2.7.13 (default, Jan 19 2017, 14:48:08)
[GCC 6.3.0 20170118] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import nacl
>>> nacl.__file__
'/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/__init__.pyc'
>>>
```

Then installed paramiko==2.2.1
```
(env-brownpapersession) bastb@bastb-vps:/var/www/brownpapersession/dev/brownpapersession$ pip install -U paramiko
Collecting paramiko
  Using cached paramiko-2.2.1-py2.py3-none-any.whl
Requirement already up-to-date: pynacl>=1.0.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: pyasn1>=0.1.7 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: bcrypt>=3.1.3 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: cryptography>=1.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: six in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from pynacl>=1.0.1->paramiko)
Requirement already up-to-date: cffi>=1.4.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from pynacl>=1.0.1->paramiko)
Requirement already up-to-date: ipaddress in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: idna>=2.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: asn1crypto>=0.21.0 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: enum34 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: pycparser in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cffi>=1.4.1->pynacl>=1.0.1->paramiko)
Installing collected packages: paramiko
  Found existing installation: paramiko 2.1.3
    Uninstalling paramiko-2.1.3:
      Successfully uninstalled paramiko-2.1.3
Successfully installed paramiko-2.2.1
```

I then ran `cat </dev/random >/dev/null` as suggested by @ploxiln and attempted starting `python manage.py shell` from another shell:
```
20:25:02 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core", 0x7ffe333b6390) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.x86_64-linux-gnu.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_coremodule.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.py", O_RDONLY) = 16
20:25:02 fstat(16, {st_mode=S_IFREG|0644, st_size=1079, ...}) = 0
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.pyc", O_RDONLY) = 17
20:25:02 fstat(17, {st_mode=S_IFREG|0644, st_size=1203, ...}) = 0
20:25:02 read(17, "\3\363\r\n\251\4\336Xc\0\0\0\0\0\0\0\0\2\0\0\0@`\1\0sh\0\0\0d\0"..., 4096) = 1203
20:25:02 fstat(17, {st_mode=S_IFREG|0644, st_size=1203, ...}) = 0
20:25:02 read(17, "", 4096)             = 0
20:25:02 close(17)                      = 0
20:25:02 close(16)                      = 0
20:25:02 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils", 0x7ffe333b6390) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.x86_64-linux-gnu.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utilsmodule.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.py", O_RDONLY) = 16
20:25:02 fstat(16, {st_mode=S_IFREG|0644, st_size=1311, ...}) = 0
20:25:02 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.pyc", O_RDONLY) = 17
20:25:02 fstat(17, {st_mode=S_IFREG|0644, st_size=1191, ...}) = 0
20:25:02 read(17, "\3\363\r\n\251\4\336Xc\0\0\0\0\0\0\0\0\2\0\0\0@`\1\0s^\0\0\0d\0"..., 4096) = 1191
20:25:02 fstat(17, {st_mode=S_IFREG|0644, st_size=1191, ...}) = 0
20:25:02 read(17, "", 4096)             = 0
20:25:02 close(17)                      = 0
20:25:02 close(16)                      = 0
20:25:02 getrandom("\25\262\272`\20\323\214M\374\244\253\224B-\255\232", 16, 0) = 16
20:25:02 getrandom(",\341D\31S\2420~\332\363u\316bv\210]", 16, 0) = 16
20:25:02 close(15)                      = 0
```

Looks good; no delays there. Seeing if I would be able to get those delays using `cat`:
```
(env-brownpapersession) bastb@bastb-vps:/var/www/brownpapersession/dev/brownpapersession$ pip install -U paramiko
Collecting paramiko
  Using cached paramiko-2.2.1-py2.py3-none-any.whl
Collecting pynacl>=1.0.1 (from paramiko)
  Using cached PyNaCl-1.1.2-cp27-cp27mu-manylinux1_x86_64.whl
Requirement already up-to-date: pyasn1>=0.1.7 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: bcrypt>=3.1.3 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: cryptography>=1.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from paramiko)
Requirement already up-to-date: six in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from pynacl>=1.0.1->paramiko)
Requirement already up-to-date: cffi>=1.4.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from pynacl>=1.0.1->paramiko)
Requirement already up-to-date: ipaddress in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: idna>=2.1 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: asn1crypto>=0.21.0 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: enum34 in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cryptography>=1.1->paramiko)
Requirement already up-to-date: pycparser in /var/www/brownpapersession/dev/env-brownpapersession/lib/python2.7/site-packages (from cffi>=1.4.1->pynacl>=1.0.1->paramiko)
Installing collected packages: pynacl, paramiko
  Found existing installation: paramiko 2.1.3
    Uninstalling paramiko-2.1.3:
      Successfully uninstalled paramiko-2.1.3
Successfully installed paramiko-2.2.1 pynacl-1.1.2
```
Started `cat </dev/random >/dev/null` and tried running `python manage.py shell`. I had the `cat` command running for ~90 seconds before starting that command to learn that the issue is there again:
```
20:30:59 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core", 0x7ffc47f4c0b0) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.x86_64-linux-gnu.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_coremodule.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.py", O_RDONLY) = 16
20:30:59 fstat(16, {st_mode=S_IFREG|0644, st_size=1079, ...}) = 0
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/sodium_core.pyc", O_RDONLY) = 17
20:30:59 fstat(17, {st_mode=S_IFREG|0644, st_size=1047, ...}) = 0
20:30:59 read(17, "\3\363\r\nr<vYc\0\0\0\0\0\0\0\0\2\0\0\0@`\1\0sh\0\0\0d\0"..., 4096) = 1047
20:30:59 fstat(17, {st_mode=S_IFREG|0644, st_size=1047, ...}) = 0
20:30:59 read(17, "", 4096)             = 0
20:30:59 close(17)                      = 0
20:30:59 close(16)                      = 0
20:30:59 stat("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils", 0x7ffc47f4c0b0) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.x86_64-linux-gnu.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utilsmodule.so", O_RDONLY) = -1 ENOENT (No such file or directory)
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.py", O_RDONLY) = 16
20:30:59 fstat(16, {st_mode=S_IFREG|0644, st_size=1311, ...}) = 0
20:30:59 open("/var/www/brownpapersession/dev/env-brownpapersession/local/lib/python2.7/site-packages/nacl/bindings/utils.pyc", O_RDONLY) = 17
20:30:59 fstat(17, {st_mode=S_IFREG|0644, st_size=1087, ...}) = 0
20:30:59 read(17, "\3\363\r\nr<vYc\0\0\0\0\0\0\0\0\2\0\0\0@`\1\0s^\0\0\0d\0"..., 4096) = 1087
20:30:59 fstat(17, {st_mode=S_IFREG|0644, st_size=1087, ...}) = 0
20:30:59 read(17, "", 4096)             = 0
20:30:59 close(17)                      = 0
20:30:59 close(16)                      = 0
20:30:59 open("/dev/random", O_RDONLY)  = 16
20:30:59 poll([{fd=16, events=POLLIN}], 1, -1) = 1 ([{fd=16, revents=POLLIN}])
20:31:50 close(16)                      = 0
20:31:50 open("/dev/urandom", O_RDONLY) = 16
20:31:50 fstat(16, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 9), ...}) = 0
```

Moving up to paramiko==2.2.1 using no-binary pynacl. Will look at the `Segmentation fault` issue later on as it seems unrelated.
